File: 078. 子集.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        self.res = []

        def backtrack(path, start_index):
            self.res.append(path[:])
            for i in range(start_index, len(nums)):
                path.append(nums[i])
                backtrack(path, i + 1)
                path.pop()

        backtrack([], 0)
        return self.res

File: 078. 子集.go
package main

func subsets(nums []int) [][]int {
	res := make([][]int, 0)
	path := make([]int, 0)

	var backtrack func(startIdx int)
	backtrack = func(startIdx int) {
		// 不用写 if startIdx >= len(nums) { return }
		// 如果越界了，for 循环不会执行

		temp := make([]int, len(path))
		copy(temp, path)
		res = append(res, temp)

		for i := startIdx; i < len(nums); i++ {
			path = append(path, nums[i])
			backtrack(i + 1) // 不是 startIdx + 1
			path = path[:len(path)-1]
		}
	}

	backtrack(0)
	return res
}
